evaluate privacy

fchen365 opened the pull request:
- evaluate the privacy for one/multiple bloom filters
- test for function results.

Commit:
evaluate privacy

- evaluate the privacy for one/multiple bloom filters
- test for function results.

Commit:
Merge branch 'master' into evaluate-privacy

Commit:
license and main

- add license information
- put all executable into main()

Commit:
add num_sampling parameter

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  """
+  k = np.zeros(depth)
+  for i in range(depth): 
+    if i%2==0:
```
Put spaces between operators.

fchen365 replied:
Definitely.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+      y_d = n_bit - y_d
+      y_x = n_bit - y_x
+
+  if (type == "asymp"): ## normal approximation
+    b = 2 * np.log(q / p)
+    l = -b * y_d / n_bit + b ** 2 / 2 * y_d * (n_bit - y_d) * (2 * y_x - n_bit) / n_bit ** 3
+    l = np.exp(l)
+  elif (type == "exact"): ## exact form
+    l = ratio_of_hypergeometric_mgf(y_d=y_d, y_x=y_x, p=p, n_bit=n_bit, depth=depth) 
```
What if users input `type` other than "asymp" or "exact"? Better check `type` is in these two values, and raise error early.

fchen365 replied:
Sure, I will add a check for this.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  """
+  ## process the inputs
+  q = 1 - p
+  if y_x == None:
```
Comparison to None should be 'expr is None' . I.e., if y_x is None.

fchen365 replied:
Thanks for catching this.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  return (n_bit - y_x) / np.maximum(n_bit - y_x - y_d, 1) * l
+
+
+def evaluate_eps_given_yd(y_d, p, n_bit, type="exact", y_x=None, depth=1000):
```
evaluate_eps_given_observed_bloom_filter, evaluate epx_given_observed_one_counts? yd doesn't seem to be self-explained.

fchen365 replied:
Thanks, I will change the function names.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  lower = 0
+  upper = n_bit
+  while lower + 1 < upper: 
+    y_x = (lower + upper) // 2
```
Better to initialize y_x before the while loop. Otherwise, if you put n_bit = 1, then y_x would never be reached.

fchen365 replied:
Will do!

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  return np.log(privacy_loss)
+
+
+def find_yx_given_yd_Is_1(p, n_bit, d):
```
yx and yd are not clear. Is it possible to make it more explicit?

fchen365 replied:
Thanks, I will change the function names.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  return res
+
+
+def find_p_given_yd_is_1(e, n_bit, d=1e-4):
```
Something like find_flip_prob_given_observed_count?

fchen365 replied:
Thanks, I will change the function names.

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  lower = 0
+  upper = .5
+  while lower + 1e-7 < upper:
+    p = (lower + upper) / 2
```
Also initiate p before the while loop.

fchen365 replied:
Yep!

huangxichen1 commented on buffling/privacy.py:
```diff
@@ -0,0 +1,265 @@
+  while lower + 1e-7 < upper:
+    p = (lower + upper) / 2
+    y_x = find_yx_given_yd_Is_1(p, n_bit, d)
+    epsilon = np.array((evaluate_eps_given_yd(1, p, n_bit, depth=1000, y_x=y_x[0]), 
```
The indent doesn't seem to conform with conventions.
Either
```
a = b(c=1,
          d=2)
```
Or
```
a = b(
  c=1,
  d=2
)
```

fchen365 replied:
This make sense.

huangxichen1 commented on examples/test.py:
```diff
@@ -0,0 +1,49 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
What does this file do? Better to add docstring at the beginning of this file to explain.

fchen365 replied:
Thanks for catching.

huangxichen1 commented on examples/test.py:
```diff
@@ -0,0 +1,49 @@
+
+def main():
+  # required p for one BF
+  print(f"epsilon = {EPSILON}, delta = {DELTA}\n")
```
Print out information when running the code is nice. Do you prefer to use logging of absl? The logging module is powerful, and allows you to log your output in other format. For example, you can choose to print out the message in commandline, and/or save to a designated file location, etc.

An example: (line 17)
https://github.com/world-federation-of-advertisers/cardinality_estimation_evaluation_framework/blob/master/run_tests.py

fchen365 replied:
absl sounds great! I will use it instead.

huangxichen1 reviewed (changes requested):
Can you please add a unit testing for the privacy.py? Also, the file name is not explicit, maybe change to evaluate_shuffling_flipping_privacy.py?

Commit:
format

- add spaces around operator
- check input "type"
- initialize all parameters in binary search
- change function names
- change variable names

Commit:
add simulation

add simulation for one bloom filter

Commit:
clear up title

Commit:
explain functions

Commit:
Merge branch 'design-doc' into evaluate-privacy

Commit:
use absl for test examples

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
```
simulated -> estimated

fchen365 replied:
sure

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+
+def ratio_of_hypergeometric_mgf(
+  count_input_ones, count_output_ones, p, n_bit, depth=1000):
+  """Compute the ratio of two hypergeometric moment-generating functions.
```
Is there a **doc** that explains why ratio_of_hypergeometric_mgf can be computed via gauss_continued_fraction? You may put the formulas in a doc and add a link here. If you still have time, may further move the doc to a markdown in this github package.

fchen365 replied:
I had a doc of technical details (docs/technical-details.pdf) in pull request #10 , please have a look.

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+  return 1 / f[0]
+
+
+def ratio_of_hypergeometric_mgf(
```
Better to have some tests:
When is your algorithm guaranteed to produce the true ratio_of_hypergeometric_mgf. Any condition? How large is the error?

fchen365 replied:
Also taken care in the doc in #10

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+      Will be Passed to evaluate_gauss_continued_fraction().
+
+  Returns:
+    A real number of the simulated epsilon.
```
estimated

fchen365 replied:
sure

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+  count_input_ones, p, n_bit, d=1e-4, n_simu=50000, depth=1000):
+  """Evaluate the privacy of one bloom filter.
+
+  Evaluate the privacy of one bloom filter given the count of input ones.
```
"Empirically evaluate ..."

This sentence is enough for one-line docstring. Maybe we don't need the previous line?

fchen365 replied:
Done

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+      count_input_ones[i], count_output_ones[i], 
+      p=p, n_bit=n_bit, depth=depth) 
+  privacy_loss = p / q / r ## Pr(A(D')=X) / Pr(A(D)=X)
+  e = np.quantile(np.abs(np.log(privacy_loss)), 1 - d)
```
This means input = (y_d -> y_{d+1}).
Do you want to further estimate the privacy loss of (y_{d+1} -> y_d)?

A design doc would be super helpful. Otherwise people stop trusting you once they see such confusing parts. We don't worry much when something is wrong. We worry much when we can't tell if something is right or wrong.

fchen365 replied:
I added a design doc to pull request #10. the file name is technical-details.pdf. Please check

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+
+def estimate_privacy_of_bloom_filter(
+  n_pub, n_bit, p, d=1e-4, n_simu=100000):
+  """Calculate the privacy parameter (epsilon) of bloom filters. 
```
We may always call privacy parameter as "privacy loss function" later, following the literature.

fchen365 replied:
Both terms are used in the literature. (I think they are different things, but please correct me if I was wrong.)

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+  return e
+
+
+def estimate_privacy_of_bloom_filter(
```
this function name is a bit too general. Could you find a more specific name?

fchen365 replied:
how about "estimate_privacy_of_shuffling_blipping_of_bloom_filter"

jiayu-google commented on buffling/privacy.py:
```diff
@@ -0,0 +1,190 @@
+def estimate_privacy_of_bloom_filter(
+  n_pub, n_bit, p, d=1e-4, n_simu=100000):
+  """Calculate the privacy parameter (epsilon) of bloom filters. 
+
```
We need to clarify somewhere that you are computing the privacy loss with y_d = 0 or k, which are the worst cases.

fchen365 replied:
This function calculate epsilon for multiple BFs, including 1. There is no y_d here. 
y_d is renamed as "count_input_ones" now, thanks to Xichen's suggestion.

Commit:
add code comments

Commit:
rename functions

jiayu-google reviewed (approved):
Excellent!

huangxichen1 commented on examples/test.py:
```diff
@@ -0,0 +1,61 @@
+
+  # required p for one BF
+  data = pd.DataFrame(columns=['# input 1s', 'epsilon'])
+  for count_input_ones in (1, 12509, 37509, 62509, 87509, 99999):
```
Maybe [1] + [int(FLAGS.n_bits * i / 8) for i in [1, 3, 5, 7]] + [FLAGS.n_bits - 1]

fchen365 replied:
good point

huangxichen1 commented on examples/test.py:
```diff
@@ -0,0 +1,61 @@
+  for count_input_ones in (1, 12509, 37509, 62509, 87509, 99999):
+    p_hat = bp.evaluate_privacy_for_one_bloom_filter(
+      count_input_ones=count_input_ones, p=0.05, n_bit=FLAGS.n_bit, 
+      d=0.001, n_simu=5000)
```
Create another FLAG for n_simu.

fchen365 replied:
will do

Commit:
add assertions

Commit:
Merge branch 'master' into evaluate-privacy

Commit:
add test file

Commit:
make test

fchen365 committed:
Merge branch 'master' into evaluate-privacy
